### PR TITLE
DOC: GSoC 2024 final blog/report

### PR DESCRIPTION
**Pull request by itellaetxe**
## Summary

Adds the final blog/report for my GSoC2024 journey, and the necessary images

### CODE REVIEW

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+-  **Repository:** `TractoEncoder GSoC <https://github.com/itellaetxe/tractoencoder_gsoc>`_
+
+
+Abstract
+--------
+
+The objective of the project is to generate synthetic human tractograms with tuneable age and neurodegeneration status properties, using deep learning models. The project is inspired by the `GESTA <https://doi.org/10.1016/j.media.2023.102761>`_ paper, which proposes a generative method using an AutoEncoder architecture for unconditional tractogram generation, combined with a robust sampling strategy. In our case, we replicated the results of the GESTA paper AutoEncoder in the FiberCup dataset using TensorFlow2 as the deep learning backend, instead of the original PyTorch backend. We also investigated how to condition the AutoEncoder architecture on scalar and categorical variables to support conditional tractogram generation on age and neurodegeneration status. Finally, we also explored the use of Variational AutoEncoders, Regression AutoEncoders, and Adversarial AutoEncoders for tractogram generation under the mentioned conditions.
```
Comment (jhlegarreta) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: In our case, we replicated the results of the GESTA paper AutoEncoder in the FiberCup dataset using *TensorFlow2 as the deep learning backend, instead of the original PyTorch backend.*

I would reword this: we did not replicate results. Not sure how you performed the generated of streamlines with different lengths in https://github.com/dipy/dipy.org/pull/63/files, but I do not recall having discussed actually having reached the point where the work on the AE would enable us to move to the sampling part.

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+Proposed Objectives
+-------------------
+
+* Replicate the results of the AutoEncoder architecture in the `GESTA <https://doi.org/10.1016/j.media.2023.102761>`_ paper using TensorFlow2 as the deep learning backend.
```
Comment (jhlegarreta) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: Cross-ref FINTA: the in GESTA is the same as in FINTA (with some other additional loss terms that we did not use here).

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+  * The `FINTA <https://10.1016/j.media.2021.102126>`_ and `GESTA <https://doi.org/10.1016/j.media.2023.102761>`_ papers because they provide a relatively simple AE architecture with an open source sampling algorithm, easy to reuse.
+
+  * `Variational AutoEncoders for Regression <https://doi.org/10.1007/978-3-030-32245-8_91>`_, which provided a good starting point for conditional Variational AutoEncoders with direct application to brain aging, related to the project's objectives.
+
+  * `Attribute-based Regularization of latent spaces for variational AutoEncoders <https://doi.org/10.1007/s00521-020-05270-2>`_ and `Adversarial AutoEncoders <http://arxiv.org/abs/1511.05644>`_, which inspired the use of Attribute-based regularization for conditioning on a continuous variable (age); and the use of adversarial training for conditioning on categoric variables (fiber bundle or neurodegeneration status), respectively.
+
+* **Replicated the** `GESTA <https://doi.org/10.1016/j.media.2023.102761>`_ **architecture originally implemented in PyTorch (vanilla AutoEncoder, not variational) using TensorFlow2+Keras.**
```
Comment (jhlegarreta) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: Same as above: we replicated the FINTA AE.

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  * The upsampling layers of the Decoder block use linear interpolation in PyTorch by default, whereas in TensorFlow2, there is no native implementation for this, and nearest neighbor (NN) interpolation is used instead. This is a significant difference in implementations, and to replicate the PyTorch behavior a custom linear interpolating upsampling layer was implemented in TF2. However, after training with both NN and linear interpolation, the results were very similar, so the custom layer was not used in the final implementation.
+  * The figure below shows a summary of the replication results:
+
+    .. image:: https://github.com/dipy/dipy.org/blob/master/_static/images/gsoc/2024/inigo/fibercup_replicated.png
+      :alt: Replication of the GESTA architecture results on the FiberCup dataset.
+      :align: center
+      :width: 800
+
+
+* **Implemented a Variational AutoEncoder (VAE) architecture based on the** `GESTA <https://doi.org/10.1016/j.media.2023.102761>`_ **architecture.**
```
Comment (jhlegarreta) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: Same as above: are we sure this adds the additional loss terms? If not, this would again be FINTA. I do not remember having discussed about this.

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 2 above)
Comment (itellaetxe) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: True, changing it to FINTA instead. I mix them up every now and then, sorry...

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 4 above)
Comment (itellaetxe) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: True, I wanted to say FINTA, again.

I would say that the majority of the architecture is the same as FINTA. I am only adding 2 things to give FINTA the variational regularization:
- Two dense layers that output $\mu_z$ and $log{\sigma^2_z}$
- The KL divergence loss component

Thus, I would say that it is based on FINTA, even if it is slightly modified

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 1 above)
Comment (itellaetxe) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: We replicated the FINTA AutoEncoder, that is what I wanted to say. I mixed it up sorry. No GESTA sampling

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 4 above)
Comment (jhlegarreta) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: :+1:.

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+  * The model is found in the ``vae_model.py`` module of the `TractoEncoder GSoC <https://github.com/itellaetxe/tractoencoder_gsoc>`_ repository. The architecture can be summarized in the figure below:
+
+  .. image:: https://github.com/dipy/dipy.org/blob/master/_static/images/gsoc/2024/inigo/inigo_variational_autoencoder.png
+    :alt: VAE architecture diagram.
+    :align: center
+    :width: 800
+
+  * It was necessary to implement Batch Normalization after the Encoder convolutional layers and exponential operator clipping, to prevent gradient explosion while training the model. The figure above displays that the encoder outputs the variance of the latent space vector (::math:`\sigma^2`), though it is a common practice to output the log-variance instead. This leads to using ::math:`\sigma = e^{\frac{1}{2}\text{log}{\sigma^2}}`, which can get numerically quite unstable as it can shoot up very quickly when a sufficiently big log-variance is input. This modification allowed stable training and it was a major contribution to the robustness of the architecture. Shoutout to my lab colleague `Jorge <https://github.com/jgarciacondado>`_ for the ideas and discussions on the topic.
```
Comment (pjsjongsung) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: Some errors in markdown? When I looked at the compiled file it said \sigma instead of $\\sigma$

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+* Investigate the use of Variational AutoEncoders (VAE) for unconditional tractogram generation based on the GESTA architecture.
+* Investigate the use of Conditional Variational AutoEncoders (CVAE) for conditional tractogram generation based on the GESTA architecture.
+* Investigate the possibility to condition the tractogram generation on the fiber bundle for additional control over the process and the data generation, using Adversarial AutoEncoders with attribute-based regularization.
+
+Objectives Completed
+--------------------
+
+* **Literature Review on synthetic tractography generation using AutoEncoders**. Main inspirations: 
+
+  * The `FINTA <https://10.1016/j.media.2021.102126>`_ and `FINTA <https://doi.org/10.1016/j.media.2021.102126>`_ papers because they provide a relatively simple AE architecture with an open source sampling algorithm, easy to reuse.
```
Comment (pjsjongsung) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst:107]: One should be GESTA right? And the first link seems to be wrong.

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+  * Weighing the Kullback-Leibler loss component was also implemented, based on the `Beta-VAE <https://openreview.net/forum?id=Sy2fzU9gl>`_ work, aiming for a stronger disentanglement of the latent space. However, this parameter was never explored (it was always set to 1.0) due to its trade-off with the reconstruction accuracy, but it is a potential improvement for future work in the context of hyperparameter optimization.
+  * The figure below shows a summary of the VAE results, also using the FiberCup dataset:
+
+    .. image:: https://github.com/dipy/dipy.org/blob/master/_static/images/gsoc/2024/inigo/vanilla_vae_120_epoch_results.png
```
Comment (pjsjongsung) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst:148]: It could be better to say 'Training Data' instead of 'Original Data' (since the next figure says reconstruction of training data). Also, instead of showing #1 and #2 reconstruction, it might be better to show the original #1 and the reconstructed #1. Hopefully it is not too late to edit the figure?

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+  * Since the generator block is trying to predict the whole latent vector ::math:`z` from a single number (::math:`r`), the model was going to probably have trouble getting the necessary geometrical variability, and overal problems for generating different things from the same number. The model was not designed for generating samples and rather to regress their associated conditioning variable, and this was a major issue that had not been foreseen. This was a good lesson to learn, and it was a good opportunity to think about the importance of the generative process in the model design, and how it should be aligned with the model's objectives. The figure below shows a set of generated samples of lengths 30 and 300 (left and right), seeing that the model was generating a very constant shape, only scaled in length:
```
Comment (pjsjongsung) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: overal -> overall

**Comment by pjsjongsung** [comment]
Looks great overall! Left a few comments.

Just to make sure it would be nice to run a spell checker before submitting the final version

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 5 above)
Comment (itellaetxe) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: Remove double colon substitute for single colon -> :math:

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+* Integrate the models into DIPY.
+
+Pull Requests
+-------------
+Because the work is still in progress, no integration of the final version was done yet. However, a draft Pull Request was created to show the progress of the work in the main DIPY repository. For now, it contains the first validated AE, together with tests for the model class. `LINK TO THE PR <https://github.com/dipy/dipy/pull/3328>`_
+
+Timeline
+--------
+
++------------+-------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------+
```
Comment (deka27) [inline gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: You could implement list-table directive, will give u less trouble. Have a look here 

https://github.com/dipy/dipy.org/pull/66/files#diff-b5c486bccd768b5473dd16ece90b3a9026925ed7ac5a82faac5c2e1123394ad6R422

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 7 above)
Comment (itellaetxe) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst:148]: I agree, thanks for the suggestion. I will edit the figure because I saved it as a draw.io file, so I can easily change it :)

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 9 above)
Comment (itellaetxe) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst]: Thank you so much for this! Again, I suck at formatting and the table gave me a trillion headaches, so I appreciate it

File: gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst (hunk 6 above)
Comment (itellaetxe) [reply gsoc_final_post/inigo_tellaetxe_gsoc_2024_final_report.rst:107]: There should only be one link to FINTA, we did not dive into GESTA.

**Comment by itellaetxe** [comment]
Visualization and last changes can be seen in https://github.com/dipy/dipy.org/pull/67. 
When merged, I will put it here.
